### ISSUE TITLE
Serialization of list of all same numpy arrays blows up

### ISSUE THREAD

**Issue by mitar** [ray-project/ray#755]
It seems Ray does not handle serialization of an array which points to same numpy arrays well. I would hope that it would store the numpy array only once, and when reading, reconstruct the references back, to again keep only one numpy array.

The issue we are having is that we have a format where we read images and then have boundaries on image to extract subimages from. So we would like to split this into two steps, one which reads all images in (and same images can exist multiple times), and then another which extracts a needed boundary.

```python
import numpy as np
import ray
ray.init()

arr = np.zeros(100000000)

%time arr_id = ray.put(arr)
%time arr_val = ray.get(arr_id)

# same reference, out of memory

arrs = [arr] * 10000000

%time arrs_id = ray.put(arrs)
%time arrs_val = ray.get(arrs_id)

# transfer ids

arrs = [arr_id] * 10000000

%time arrs_id = ray.put(arrs)
%time arrs_val = ray.get(arrs_id)

# arrs_val has IDs, not values themselves :-(
```

**Comment by robertnishihara** [comment]
Closing because this is being tracked in #319.

For now, do you need to use a list with multiple copies of the same array? Or can you just use one copy of the array? In your application, are the arrays in the list *identical*? Or are they many different crops of the same underlying base array?

If you really need multiple copies then one workaround for now could be to pickle and unpickle the list of arrays, although that can be inefficient.

**Comment by pcmoritz** [comment]
So there are 2 approaches to deal with this:

As a quick workaround as you hint at, you can store the array once and then use object IDs to refer to it (object IDs can be stored inside of for example lists in the object store).

The other approach is to do this as part of numbuf transparently. Arrow now has DictionaryArrays that can be used, see

https://github.com/apache/arrow/blob/cb5e7b6fa7d75e14e163ce43cb333b02e9fe1c03/python/doc/source/data.rst

I'm planning to do this when rewriting numbuf using Cython and the pyarrow APIs, but this is probably ~1 month down the road (potentially longer). We could hack around it in the current codebase, but that's probably not worth it and it will be much easier once everything is in Cython.

**Comment by mitar** [comment]
> In your application, are the arrays in the list identical?

Yes, they are literally the same array, same reference to the same array.

BTW, in above example, putting `arrs = [arr_id] * 10000000` is also very slow, it takes like 30 seconds. So I would guess that the same mechanism mentioned in #319, which would track which object you encountered already, could also track which IDs you encountered and do not reprocess them again.

**Comment by robertnishihara** [comment]
Yeah, interesting, custom objects seem to be a bit slower than primitive types (presumably since it has to go back to Python, whereas for a list of ints it never has to leave C++).

E.g., it's slightly faster if you do

```python
x_id = ray.put(1)

%time ray.put(10000000 * [x_id.id()])  # 1.5s

%time ray.put(10000000 * [x_id])  # 26s
```

You can reproduce the ID out of the byte string since `x_id == ray.local_scheduler.ObjectID(x_id.id())`, although this is a hack and not part of the API.

But yes, this will all be handled much more nicely once we do it properly.

**Comment by mitar** [comment]
> But yes, this will all be handled much more nicely once we do it properly.

Which issue describes what should be done for this?

**Comment by robertnishihara** [comment]
I'll comment with some thoughts in #319, but there's no description of a solution there yet.
